File: matlab/dlapy2.m
function [tau] = dlapy2(x, y)
% computes sqrt(x^2 + y^2) without causing overflow.

xabs = abs(x);
yabs = abs(y);
w = max(xabs, yabs);
z = min(xabs, yabs);
if z == 0.0
    tau = w;
else
    tau = w * sqrt(1.0 + (z / w)^2);
end

File: c/dlapy2.c
#include <math.h>
#include "dstedc.h"

double dlapy2(double x, double y)
// computes sqrt(x^2 + y^2) without causing overflow.
{
    double xabs = fabs(x);
    double yabs = fabs(y);
    double w = fmax(xabs, yabs);
    double z = fmin(xabs, yabs);
    double tau;

    if (z == 0.0)
        tau = w;
    else
        tau = w * sqrt(1.0 + SQ(z / w));

    return tau;
}
